File: Java/935.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {

    static int[] res;
    public int knightDialer(int N) {
        if (res == null) {
            res = new int[5001];
            res[1] = 10;
            int mod = 1000_000_007;
            long f0 = 1, f1379 = 4, f28 = 2, f46 = 2;
            for (int i = 2; i < res.length; ++i) {
                long b0 = f46;
                long b1379 = (f28 << 1) + (f46 << 1);
                long b28 = f1379;
                long b46 = (f0 << 1) + f1379;
                f0 = b0 % mod;
                f1379 = b1379 % mod;
                f28 = b28 % mod;
                f46 = b46 % mod;
                res[i] = (int)((f0 + f1379 + f28 + f46) % mod);
            }
        }
        return res[N];
    }
}
__________________________________________________________________________________________________
sample 31764 kb submission
class Solution {
    public int knightDialer(int N) {
        if(N==1)
            return 10;
        
        //initialization
        long c46=2,c1379=4,c28=2,c0=1;
        int mod=(int)1e9+7;
        int i=2;
        
        while(i<=N){
            //hop
            //key 0 <-- key 4,6
            //key 2,8 <-- key 1,3,7,9
            //...
            long n0=c46,n28=c1379,n1379=(c46*2+c28*2)%mod,n46=(c1379+c0*2)%mod;
            
            c0=n0;
            c28=n28;
            c46=n46;
            c1379=n1379;
            
            i++;
        }
        
        return (int)((c0+c28+c46+c1379)%mod);
    }
}
__________________________________________________________________________________________________

File: Python3/935.py
__________________________________________________________________________________________________
sample 60 ms submission
cache = [[1]*10]
moves = [[4,6],[6,8],[7,9],[4,8],[0,3,9],[],[0,1,7],[2,6],[1,3],[2,4]]
mod = 10**9 + 7

class Solution:
    def knightDialer(self, N: int) -> int:
        while len(cache) < N:
            dp = [0]*10
            for d,s in enumerate(cache[-1]):
                for nei in moves[d]:
                    dp[nei] += s
            cache.append(dp)
        return sum(cache[N-1])%mod
__________________________________________________________________________________________________
sample 12936 kb submission
class Solution:
    def knightDialer(self, N: int) -> int:
        allowable = {1: [6, 8],
             2: [7, 9],
             3: [4, 8],
             4: [3, 9, 0],
             5: [],
             6: [1, 7, 0],
             7: [2, 6],
             8: [1, 3],
             9: [4, 2],
             0: [4, 6]}

#         dp = [[0] * N for i in range(10)]
        dp = [1] * 10
        for j in range(N - 1):
            dp2 = [0] * 10
            for i in range(10):
                for nei in allowable[i]:
                    dp2[i] += dp[nei]
            dp = dp2       
        return sum(dp) % (10 ** 9 + 7)
__________________________________________________________________________________________________
